### PR TITLE
Suspendable request helpers

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
Based on #8, #19, #22, #17. Blocked until those PRs are merged.

**Comment by gavin-norman-sociomantic** [comment]
Prerequisites merged. Unblocked.

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 1]
```diff
@@ -0,0 +1,1165 @@
...
+public struct SuspendableRequestInitialiser ( Request, FillPayload,
+    HandleStatusCode )
+{
+    import swarm.neo.client.RequestOnConn;
+    import swarm.neo.connection.RequestOnConnBase;
+    import swarm.neo.client.mixins.AllNodesRequestCore : AllNodesRequestInitialiser;
+
+    /// Struct encapsulating the logic for inc/decrementing the request's
+    /// SuspendableRequestSharedWorkingData.ready_for_state_change counter.
+    private struct StateChangeReadinessTracker
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d:195]: I like encapsulation of concerns this struct is doing.

**Comment by gavin-norman-sociomantic** [comment]
Added release notes.

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 2]
```diff
@@ -0,0 +1,1165 @@
...
+        Called by AllNodesRequest to reset any state required for initialisation
+        to begin again. Decrements the
+        SuspendableRequestSharedWorkingData.ready_for_state_change counter, if
+        it has been incremented for this request-on-conn.
+
+    ***************************************************************************/
+
+    public void reset ( )
+    {
+        this.ready_for_state_change = false;
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: `this.all_nodes_initialiser.reset()` for future-proofness (as it is empty right now, might not be in the future)?

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 3]
```diff
@@ -0,0 +1,1165 @@
...
+
+    Returns:
+        instance of AllNodesRequest constructed with the provided arguments
+
+*******************************************************************************/
+
+public AllNodesRequest!(Request, Connector, Disconnected,
+    SuspendableRequestInitialiser!(Request, FillPayload, HandleStatusCode),
+    Handler)
+    createSuspendableRequest ( Request, Connector, Disconnected, FillPayload,
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: I have no better suggestion here, but it's very hard to spot the function's name with the current indentation (all these helper functions have the same problem, but this one is the worst).

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 4]
```diff
@@ -0,0 +1,1165 @@
...
+        ***********************************************************************/
+
+        public bool stop ( )
+        {
+            return this.changeDesiredState(MessageType.Stop);
+        }
+
+        /***********************************************************************
+
+            Changes the desired state to that specified. Sets the desired state
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d:582]: `to that specified` or `to the specified`?

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 5]
```diff
@@ -0,0 +1,1165 @@
...
+            flag and resumes any handler fibers which are suspended, passing the
+            control message flag to the fiber via the return value of suspend().
+
+            If one or more connections are not ready to change state, the
+            control change doesn not occur. A connection is ready to change the
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: typo `doesn not`

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 6]
```diff
@@ -0,0 +1,1165 @@
...
+            }
+
+            // Send state change message to node and wait for ACK.
+            stopped = this.changeRequestState();
+
+            // Notify the user that the state change has taken effect and handle
+            // the case where they request a new state change in the notifier.
+            state_changed_in_notifier = this.notifyUser(stopped);
+        }
+        while ( !stopped );
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: I have no words how good this is.

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 7]
```diff
@@ -0,0 +1,1165 @@
...
+        Params:
+            stopped = indicates whether the just ACKed state change was to stop
+                the request (true) or was another type of state change (false)
+
+        Returns:
+            true if another state change was requested during notification
+
+    ***************************************************************************/
+
+    private bool notifyUser ( bool stopped )
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: I was trying to see how this flag is used, but it seems that it's not used at all (less complexity, which is good).

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 8]
```diff
@@ -0,0 +1,1165 @@
...
+        Stopped
+    }
+
+    /// Desired state of the request, as set by the user API.
+    public DesiredState desired_state = DesiredState.Running;
+
+    /***************************************************************************
+
+        The number of handlers that are currently in the process of sending a
+        control message to the node of waiting for an acknowledgement message
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: `or waiting for an acknowledgement`

**Review by nemanja-boric-sociomantic — COMMENTED** [review]
comments for the first commit

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d (hunk 4 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/SuspendableRequestCore.d:582]: "to that" is correct.

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/client/mixins/SuspendableRequestCore.d:582]: Thanks! I was unsure.

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d (hunk 6 above)
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: Hmm wait... there's some conflict between this and https://github.com/gavin-norman-sociomantic/swarm-tsunami/blob/2ae0bb629d97d29219d121f5d569a93473a72355/test/neo/client/request/internal/GetAll.d#L413. I'll have to look into this. It looks like the same logic has been duplicated in two place, in slightly different forms.

Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: Ah, ok, I haven't got to that point yet.

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d (hunk 7 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/SuspendableRequestCore.d]: Hmm wait... there's some conflict between this and https://github.com/gavin-norman-sociomantic/swarm-tsunami/blob/2ae0bb629d97d29219d121f5d569a93473a72355/test/neo/client/request/internal/GetAll.d#L413. I'll have to look into this. It looks like the same logic has been duplicated in two places, in slightly different forms. (`stopped` is used in the other module.)

**Comment by gavin-norman-sociomantic** [comment]
Updated, removing doubled logic in `SuspendableRequestControllerFiber` and the client's GetAll implementation (now simplified).

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d (hunk 3 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/SuspendableRequestCore.d]: Yeah, I'll see if I can come up with anything better.

**Comment by gavin-norman-sociomantic** [comment]
Updated:
* Call `AllNodesRequestInitialiser.reset` in `SuspendableRequestInitialiser.reset`.
* Fixed typos.
* Improved formatting of long function signatures.

**Comment by nemanja-boric-sociomantic** [comment]
Have you forgot to push?

**Comment by nemanja-boric-sociomantic** [comment]
(https://github.com/sociomantic-tsunami/swarm/pull/25/commits/5dc65393c47744638a8b2b61def00ed4dca4e5c5#diff-0ad780cddea50703f1972818dba99322R296 still doesn't call `superinitialiser.reset()` for example)

**Comment by nemanja-boric-sociomantic** [comment]
Ah no, it's in the wrong (the second) commit.

**Comment by gavin-norman-sociomantic** [comment]
>Ah no, it's in the wrong (the second) commit.

Updated.

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 9]
```diff
@@ -322,10 +322,12 @@ public struct SuspendableRequestInitialiser ( Request, FillPayload,
 
 *******************************************************************************/
 
-public SuspendableRequestInitialiser!(Request, FillPayload, HandleStatusCode)
-    createSuspendableRequestInitialiser ( Request, FillPayload, HandleStatusCode )
-    ( RequestOnConn.EventDispatcherAllNodes conn, Request.Context* context,
-    FillPayload fill_payload, HandleStatusCode handle_status_code )
+public
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: this is still in the wrong commit

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/SuspendableRequestCore.d]: Ah shit, that too.

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d [hunk 10]
```diff
@@ -580,7 +584,7 @@ public template SuspendableController ( Request, IController, MessageType )
             control message flag to the fiber via the return value of suspend().
 
             If one or more connections are not ready to change state, the
-            control change doesn not occur. A connection is ready to change the
+            control change does not occur. A connection is ready to change the
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/SuspendableRequestCore.d]: ditto for all other stuff in this file I guess.

File: src/swarm/neo/client/mixins/SuspendableRequestCore.d (hunk 9 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/SuspendableRequestCore.d]: Updated. (There was a bunch of other stuff wrongly in the second commit too!)

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Comment by nemanja-boric-sociomantic** [comment]
Now I'm excited :-).

**Pull request merged**